### PR TITLE
Table

### PR DESCRIPTION
**Pull request by elliotwoods**
As per discussion today.

I've fixed some of the naming:
- Table outputs a Table
- Table inputs are capitalized
- AsValue is now GetValues

Insert is split into Insert and Remove nodes

I didn't make the GetColumns node yet. so more to come in the future!

### CODE REVIEW

**Comment by elliotwoods** [comment]
ah sorry, i just added a fix to SpinCursor to this PR (Enabled was set to false by default)

**Comment by joreg** [comment]
now that you point at that please also fix its output pin name to "Layer" and why not spend it a color input (default white) and make it compatible with the other Cursor (DX9) that takes as mousestate as input instead of XY

**Comment by elliotwoods** [comment]
MouseState wouldn't make much sense as mouse not often linked to this cursor (mapping with a mouse? where's the nudge controls?). 
Agree on colour input
Other question is whether this belongs in Image pack or addonpack. (actually, i've got it in both at the moment by accident)

**Comment by joreg** [comment]
ahk, thought it was just a cursor alternative..what else is it actually?

also don't forget my request for category rename (see above)

**Comment by elliotwoods** [comment]
It's a cursor for mapping. 
It's designed so that you can tell where it is relative to a surface which is being projected on. (Your eye naturally can figure out the axis of the spin, and it becomes easier to navigate a cursor without needing to actually see it).

**Comment by joreg** [comment]
well then if it is so specific to mapping/image stuff i'd argue put it in the imagepack

**Comment by joreg** [comment]
bump

**Comment by elliotwoods** [comment]
i'll have a windows environment setup again at the end of this week
will check then (me in uk, shuttle is in seoul)

On 5 January 2014 14:56, joreg notifications@github.com wrote:

> bump
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/vvvv/vvvv-sdk/pull/153#issuecomment-31606015
> .

## 

## 

Elliot Woods
elliot elliot@kimchiandchips.com@KimchiAndChips.comhttp://www.kimchiandchips.com/

**Comment by joreg** [comment]
bump

**Comment by kimadactyl** [comment]
I think the table node save bang is kaput!

**Comment by joreg** [comment]
bump

**Comment by elliotwoods** [comment]
so to confirm, i just need to remove the SpinCursor for the time being?

**Comment by elliotwoods** [comment]
also i'll double check the save bang input (i was having trouble with that also, need to check if it's fixed)

also to confirm, when this is pulled will it end up in x64 and x86 addon packs? or do i have to do something else to make that happen?

**Comment by elliotwoods** [comment]
ok this looks loaded and ready

**Comment by elliotwoods** [comment]
also, don't forget to add the project the AddonPack.sln!
due to NuGet, everybody's solution gets edited
so i suggest you add on your end and merge that

**Comment by joreg** [comment]
as mentioned above:
how about setting category of the main node to "Spreads". and all the nodes using it have "Table" thats fine. as we have with Contours (CV.Image) and then Perimeter (CV.Contours)

also how about tags like "store, persistent"

**Comment by joreg** [comment]
also @woeishi didn't you mention you have this in quite a rework? whats the state of that?

**Comment by woeishi** [comment]
state is good, all working so far (string and values). just didn't have 
time to wrap my head around how to deal with tables with mixed types yet

**Comment by joreg** [comment]
hmm..is that in a public branch somewhere? can @elliotwoods  have a look at it? will it change much?

**Comment by elliotwoods** [comment]
i was considering to allow mixed types, but couldn't think of a perfectly clean way
really curious to see how it works!
is it a rework or did you start from scratch?

**Comment by woeishi** [comment]
it's based off your branch, though now the only things which are still 
kind of the same is the table view. it's spreaded tables and all the 
modifiers are generic to enable all kinds of types, once there exists a 
component for the gridview (was thinking of colour, matrix, etc)

however i didn't properly git branch remote from yours, so now it's 
somewhere outside my fork. @elliot: mind if i just upload or send a zip?

**Comment by elliotwoods** [comment]
go ahead! send a zip!
i know a lot of people are using the value Table right now in a range of projects
so i'm keep to see something compatible hit the Addonpack asap

**Comment by joreg** [comment]
@woeishi is anything (probably besides naming) speaking against the value part being compatible?

**Comment by woeishi** [comment]
depends how you define compatible
for the user it'll definitely appear to be the same.
but in order to have it all spreaded and multiple tables serialized 
into one table it will not be compatible under the hood

**Comment by joreg** [comment]
well, compatible in a way that if we merge what we have now and someone uses it and we later merge your stuff...it still works for them.

**Comment by woeishi** [comment]
the main nodes are compatible appearance-wise (Table and TableView), saved xml data is compatible
insert and remove -> insertrow/column/cell and removerow/column/cell
setrow as one less binsize pin which i didn't understand what the usage was
getvalues -> dump, since it returns the whole table

so for alpha addonpack it's compatible enough i'd say, no changes, which are not diffable in case there's a beta release soon

File: vvvv45/addonpack/src/nodes/plugins/Spreads/Table/GetValues.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+using System.Linq;
+using System.Text;
+using VVVV.PluginInterfaces.V2;
+using System.ComponentModel.Composition;
+
+namespace VVVV.Nodes.Table
+{
+	#region PluginInfo
+	[PluginInfo(Name = "GetValues", Category = "Table", Version = "Value", Help = "Convert Table to Values", Tags = "", Author = "elliotwoods", AutoEvaluate = true)]
+	#endregion PluginInfo
```
Comment (joreg) [inline vvvv45/addonpack/src/nodes/plugins/Spreads/Table/GetValues.cs:12]: does this really need AutoEvaluate=true?

**Comment by elliotwoods** [comment]
@woeishi 's version is rocking!
GetRow, GetColumn, SetRow, SetColumn, SetCell, etc

The only thing I don't 'like' about it is 'Table Name'. I think I can see why (if you're spreading the tables), but it seems a bit superfluous.

But anyway in general, I think this version hits a lot of new high notes.
+1 for implemneting @woeishi 's from the start

Note that after you pull it, I'd like to make some changes to it (e.g. with Auto Save on, changing the filename should auto-load from that file if the file exists).

elliot

File: vvvv45/addonpack/src/nodes/plugins/Spreads/Table/GetValues.cs (hunk 1 above)
Comment (elliotwoods) [inline vvvv45/addonpack/src/nodes/plugins/Spreads/Table/GetValues.cs:12]: no

**Comment by woeishi** [comment]
@elliotwoods table name is indeed for spreaded tables, serializing doesn't work without dedicated names.
it's quite convenient though, when your have a couple of different tables.

about saving: i had troubles saving with your version, but didn't investigate further due to the spreading rewrite. as far as i remember it's reproduceable when you create and set values via TableView only

also thought about moving table creation to a dedicated node instead of the main table one. since it doesn't make much sense when load from xml the table might be different to the config in the pins

**Comment by elliotwoods** [comment]
can we pull something please
either mine or woei's
mine is a little simpler, whilst woei's is a little more advanced.
i'm not 100% convinced that spreaded tables is a good thing (but that's just because i've never imagined using it), but anyway, let's pull woei's i think
(unhappy that another addonpack has come out without table. let me know what i can do!)

**Comment by woeishi** [comment]
tonfilm and i were in the middle of a huge rework, when the last beta was released...
still needs some cleaning up, but i'll make it within the next beta release, i promise.

**Comment by elliotwoods** [comment]
huge rework? sounds scary

**Comment by tebjan** [comment]
@woeishi can we consider this as closed now? and should we move it into the addonpack after all?
@elliotwoods what are your thoughts?

**Comment by woeishi** [comment]
i'd say lets close it for now. gather some feedback in the dedicated 
repo. we can still add it to the addonpack later. and it's not tied to 
vvvv versions through any complex dependency anyways

**Comment by elliotwoods** [comment]
i seriously think this (your version) needs to make it into the addonpack asap
it's been in limbo for so long that it might as well not exist

**Comment by woeishi** [comment]
why do you want that in the addonpack so urgently? :D

i don't mind, would try some subtree merge
what does the dev section say?

**Comment by tebjan** [comment]
Lets close this PR and wait for some feedback on the contribution. As soon as there are enough complaints we addonpack it...

and for the reference: http://vvvv.org/contribution/vvvv.datatable

**Pull request closed without merging**